xwayland/selection/incoming.c

```diff
@@ -25,113 +40,115 @@ static int xwm_data_source_write(int fd, uint32_t mask, void *data) {
-		}
+	if (len < remainder) {
+		transfer->property_start += len;
+		return 1;
+	} else if (transfer->incr) {
+		xwm_notify_ready_for_next_incr_chunk(transfer);
+	} else {
+		wlr_log(WLR_DEBUG, "transfer complete");
+		xwm_selection_transfer_finished(transfer);
 	}
 
-	return 1;
+	return 0;
 }
 
-static void xwm_write_property(struct wlr_xwm_selection_transfer *transfer,
-		xcb_get_property_reply_t *reply) {
-	struct wlr_xwm *xwm = transfer->selection->xwm;
-
-	transfer->property_start = 0;
-	transfer->property_reply = reply;
-
-	xwm_data_source_write(transfer->wl_client_fd, WL_EVENT_WRITABLE, transfer);
+static void xwm_write_selection_property_to_wl_client(
+		struct wlr_xwm_selection_transfer *transfer) {
+	if (transfer->incr && transfer->wl_client_fd < 0) {
+		// Wayland client closed its pipe prematurely before the X11 client finished
+		// its incremental transfer. Continue draining the X11 client.
+		xwm_notify_ready_for_next_incr_chunk(transfer);
+		return;
+	}
 
-	if (transfer->property_reply != NULL) {
-		struct wl_event_loop *loop =
-			wl_display_get_event_loop(xwm->xwayland->wl_display);
-		transfer->event_source = wl_event_loop_add_fd(loop,
-			transfer->wl_client_fd, WL_EVENT_WRITABLE, xwm_data_source_write,
-			transfer);
+	bool wl_client_finished_consuming = !write_selection_property_to_wl_client(
+			transfer->wl_client_fd, WL_EVENT_WRITABLE, transfer);
+	if (wl_client_finished_consuming) {
+		return;
 	}
+
+	// Wrote out part of the property to the Wayland client, but the client was
+	// unable to accept all of it. Schedule an event to asynchronously complete
+	// the transfer.
+	struct wl_event_loop *loop =
+		wl_display_get_event_loop(transfer->selection->xwm->xwayland->wl_display);
+	transfer->event_source = wl_event_loop_add_fd(loop,
+		transfer->wl_client_fd, WL_EVENT_WRITABLE,
+		write_selection_property_to_wl_client, transfer);
 }
 
-void xwm_get_incr_chunk(struct wlr_xwm_selection_transfer *transfer) {
+static bool xwm_selection_transfer_get_selection_property(
+		struct wlr_xwm_selection_transfer *transfer, bool delete) {
 	struct wlr_xwm *xwm = transfer->selection->xwm;
-	wlr_log(WLR_DEBUG, "xwm_get_incr_chunk");
 
-	xcb_get_property_cookie_t cookie = xcb_get_property(xwm->xcb_conn,
-		0, // delete
+	xcb_get_property_cookie_t cookie = xcb_get_property(
+		xwm->xcb_conn,
+		delete,
 		transfer->selection->window,
 		xwm->atoms[WL_SELECTION],
 		XCB_GET_PROPERTY_TYPE_ANY,
 		0, // offset
 		0x1fffffff // length
 		);
 
-	xcb_get_property_reply_t *reply =
+	transfer->property_start = 0;
+	transfer->property_reply =
 		xcb_get_property_reply(xwm->xcb_conn, cookie, NULL);
-	if (reply == NULL) {
+
+	if (!transfer->property_reply) {
 		wlr_log(WLR_ERROR, "cannot get selection property");
+		return false;
+	}
+
+	return true;
+}
+
+void xwm_get_incr_chunk(struct wlr_xwm_selection_transfer *transfer) {
+	wlr_log(WLR_DEBUG, "xwm_get_incr_chunk");
+
+	if (transfer->property_reply) {
```

Comment:
Note for reviewer: I ran into this being a source of compositor segfaults while introducing bugs in early drafts of this code. Since it's possible for an X11 client to do this, I think it's worth guarding against.

Current behavior doesn't try to do anything fancy in this obscure case. An error message will be printed, and the clipboard will become "stuck" because we're no longer servicing the transfer.